bigint/src/lib.rs

```diff
@@ -1349,6 +1356,47 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = (bits + big_digit::BITS - 1) %
+                    big_digit::BITS + 1;
```

Comment:
bits shouldn't be 0 but IMO it's better and more readable to do it the way
that's correct in general, and post optimizer there's no difference anyways.

On Thu, Jul 14, 2016 at 2:12 PM, Josh Stone <notifications@github.com>
wrote:

> In bigint/src/lib.rs
> <https://github.com/rust-num/num/pull/204#discussion_r70893665>:
>
> > @@ -1349,6 +1356,47 @@ impl Integer for BigUint {
> >      }
> >  }
> >
> > +fn high_bits_to_u64(v: &BigUint) -> u64 {
> > +    match v.data.len() {
> > +        0   => 0,
> > +        1   => v.data[0] as u64,
> > +        _   => {
> > +            let mut bits = v.bits();
> > +            let mut ret = 0u64;
> > +            let mut ret_bits = 0;
> > +
> > +            for d in v.data.iter().rev() {
> > +                let digit_bits = (bits + big_digit::BITS - 1) %
> > +                    big_digit::BITS + 1;
>
> Why are you adding BITS first? That should be a modular no-op. Only
> reason I can think is if bits were 0, to avoid subtraction underflow, but
> that should never be the case AFAICT.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-num/num/pull/204/files/a1772f2a0d3e8fb7384a89322c93816b017fcc57#r70893665>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB5r7mDoQNr2wrphE-LLhn4F55qAfqLeks5qVrRCgaJpZM4JIeee>
> .
>